### ISSUE TITLE
GetCursorPos and GetWindowContentRegionMin/Max grow in opposite directions

### ISSUE THREAD

**Issue by grimdoomer** [ocornut/imgui#5048]
**Version/Branch of Dear ImGui:**

Version: 1.82 (18200)
Branch: master
Wrapper: [ImGui.Net](https://github.com/mellinoe/ImGui.NET)

**Back-end/Renderer/Compiler/OS**

Back-ends: Custom backend (based on d3d11 + win32)
Compiler: MSVC
Operating System: Windows 7

**My Issue/Question:**

I'm trying to implement clipping logic for widgets that are displayed in a window similarly to how the ImGuiListClipper works. To do this I was trying to find a correlation between GetCursorPosY() and the Y value of GetWindowContentRegionMin()/Max(). But to my surprise checking if GetCursorPosY() was between the Y value of GetWindowContentRegionMin() and GetWindowContentRegionMax() did not seem to work. Based on the comments for these functions they are all in "window space" so this seemed like the intuitive correlation until I did some debugging.

From debugging I discovered that GetCursorPosY() grows in the positive direction as widgets are drawn, while GetWindowContentRegionMin()/Max() grows in the negative direction as you scroll downward in a window.

Printing the value of GetCursorPosY() for every widget drawn (with my clipping disabled) you can see the cursor position grows in the positive Y direction:
```
Cursor Y: 46
Cursor Y: 69
Cursor Y: 171
Cursor Y: 194
Cursor Y: 396
Cursor Y: 471
Cursor Y: 494
Cursor Y: 569
...
```

In the gif attached you can see the Y values of GetWindowContentRegionMin() ("Content region min") and GetWindowContentRegionMax() ("Content region max) as I scroll downward in the window, and they grow in the negative direction. 
![output](https://user-images.githubusercontent.com/15973121/155305148-e836d38c-5316-4760-ae43-28979b10d131.gif)

I found the following comments for these functions that state the current scroll value is subtracted to produce the return value:
```
IMGUI_API ImVec2 GetWindowContentRegionMin();   // content boundaries min for the full window (roughly (0,0)-Scroll), in window coordinates
IMGUI_API ImVec2 GetWindowContentRegionMax();   // content boundaries max for the full window (roughly (0,0)+Size-Scroll) where Size can be override with SetNextWindowContentSize(), in window coordinates
```

By adding 2 * GetScrollY() to the Y value of GetWindowContentRegionMin() and GetWindowContentRegionMax() the clipping logic then worked as expected by checking if GetCursorPosY() was between the min/max content region Y values. Here is the working code I ended up with to check if a widget should be clipped or not:
```
// Get the current position and content region min/max bounds.
float currentPos = ImGui.GetCursorPosY();
float contentRegionMin = ImGui.GetWindowContentRegionMin().Y + (2f * ImGui.GetScrollY());
float contentRegionMax = ImGui.GetWindowContentRegionMax().Y + (2f * ImGui.GetScrollY());

// Calculate the end pos of the field based on its height.
float fieldEndPos = currentPos + field.FieldHeight;

// Check if this is the first frame or if the field is in view. Either the starting pos or ending pos of the field must be between [contentRegionMin, contentRegionMax], or
// the field is larger than the content region and contentRegionMin/contentRegionMax must be between [currentPos, fieldEndPos].
if (this.firstFrame == true || (currentPos <= contentRegionMax && currentPos >= contentRegionMin) || (fieldEndPos <= contentRegionMax && fieldEndPos >= contentRegionMin) || (currentPos <= contentRegionMax && fieldEndPos >= contentRegionMin))
{
	// Update the field position.
	field.FieldPosition = currentPos;

	// Draw the field.
	bool updated = this.tagFieldRenderFunctions[field.CachedFieldInfo.TagFieldType](field, owner);

	// Update field size and return field draw result.
	field.FieldHeight = Math.Abs(ImGui.GetCursorPosY() - currentPos);
	return updated;
}
else
{
	// Update the field pos and skip forward in the content pane.
	field.FieldPosition = currentPos;
	ImGui.SetCursorPosY(ImGui.GetCursorPosY() + field.FieldHeight);
}
```

I'm not sure if GetCursorPosY() and GetWindowContentRegionMin()/GetWindowContentRegionMax() growing in opposite directions is expected behavior or not? It almost seems like GetWindowContentRegionMin()/Max() should be adding the current scroll offset and not subtracting it.

**Comment by Rockstar50373** [comment]
Thank you for your email!

**Comment by ocornut** [comment]
Hello Ryan,

> I'm trying to implement clipping logic for widgets that are displayed in a window similarly to how the ImGuiListClipper works.

Your question starts with a XY Problem (https://xyproblem.info/)

Why do you need to do that?

>  Here is the working code I ended up with to check if a widget should be clipped or not:

Why not using either the list clipper, either `IsRectVisible()` which does exactly that?

Both CursorPos and GetWindowContentRegionXXX are values **relative to window position**, which is a little meh and I am am trying to get rid off in the API (best to use absolute coordinates). ContentRegion is quite bizarrely defined and will be replaced eventually, they are the boundaries used for right-alignment and bottom-alignment and can't actually take account of scrolling range (unless that range is explicitly overridden with SetNextWindowContentSize).
Instead of printing the numbers use `GetForegroundDrawList()->AddRect(region_min + window_pos, region_max + window_pos)` and you'll understand what this rectangle means (spoiler: it's weird but correct for that it needs to do). You can also see them all under Metrics->Tools.
The TL;DR; if you SHOULD NOT need to use the ContentRegion Rectangle for what you are trying to do, it's not just the adequate data type. I would always want to get back to the initial problem first. You probably need to use InnerClipRect or current ClipRect but then it is unclear what you wouldn't use existing helper functions so we can't answer you correctly without you defining your problem.

**Comment by grimdoomer** [comment]
> Why do you need to do that?

Right, so as part of my application I display editors for game data that is represented as C structs. Each field in a struct is represented by one or more ImGui widgets and some need 10 or more widgets to fully represent (e.x.: bitmasks). Some of these structs contain hundreds of individual fields, and trying to draw and run logic updates for (what ends up being) thousands of ImGui widgets tanks the fps of the application. To address this I decided to check if a field would be clipped based on the current view of the window that is visible, and skip drawing and logic updates for anything not in view.

> Why not using either the list clipper, either IsRectVisible() which does exactly that?

As far as I can tell the list clipper only works with widgets that are the same height, whereas the the widgets I'm using to represent these struct fields are all different heights, and can change dynamically based on other interactions that happen in the tool. As for IsRectVisible() I just never found this while searching around. I will give this a try in place of what I'm currently doing.

**Comment by grimdoomer** [comment]
IsRectVisible() works just great, thanks for the help.
